### ISSUE TITLE
Add metrics support to splunk_hec sink

### ISSUE THREAD

**Issue by chris1786** [vectordotdev/vector#7993]
### Current Vector Version

```
vector 0.14.0 (x86_64-unknown-linux-musl 5f3a319 2021-06-03)
```

### Use-cases

Send `counter` and `gauge` metrics to a Splunk HTTP Event Collector.

### Attempted Solutions

It is possible to post metrics to a Splunk HEC using the existing`http` sink. This involves the following transform steps:

1. Use `filter` transform to select `gauge` and `counter` metrics
2. Use `metric_to_log` to convert the metrics into logs
3. Use `remap` to munge the data into the format required by the HEC

While this works, it is somewhat clunky and unintuitive.

### Proposal

I propose to amend the `splunk_hec` sink so that it accepts metric inputs, as well as log inputs. `splunk_hec` would take in all metrics, discard the metric types that Splunk doesn't support, and transform the metrics into the format expected by the Splunk HEC.

An alternative approach would be to create an additional sink for metrics, but the above proposal feels cleaner.

**Comment by chris1786** [comment]
To add to this, if the proposal is accepted, I am happy to develop the solution and raise a PR.

**Comment by jszwedko** [comment]
@chris1786 this would be welcome! I think what you've outlined makes sense. It seems like the events just need to match a schema like:

```
{
  "time": 1486683865,
  "event": "metric",
  "source": "disk",
  "host": "host_1.splunk.com",
  "fields": {
    "region": "us-west-1",
    "datacenter": "dc1",
    "rack": "63",
    "os": "Ubuntu16.10",
    "arch": "x64",
    "team": "LON",
    "service": "6",
    "service_version": "0",
    "service_environment": "test",
    "path": "/dev/sda1",
    "fstype": "ext3",
    "metric_name:cpu.usr": 11.12,
    "metric_name:cpu.sys": 12.23,
    "metric_name:cpu.idle": 13.34
  }
}
```

https://docs.splunk.com/Documentation/Splunk/8.2.0/Metrics/GetMetricsInOther#Get_metrics_in_from_clients_over_HTTP_or_HTTPS

Is that accurate?

In addition to supporting it in the `splunk_hec` sink, it could be nice to support it in the `splunk_hec` source, but we can leave that for later.

**Comment by chris1786** [comment]
@jszwedko thanks for the reply. 

The example you've posted is supported by Splunk version 8.x.

Version 7.x supports a slightly different format:

```
{
  "time": 1486683865,
  "event": "metric",
  "source": "disk",
  "host": "host_1.splunk.com",
  "fields": {
    "region": "us-west-1",
    "datacenter": "dc1",
    "rack": "63",
    "os": "Ubuntu16.10",
    "arch": "x64",
    "team": "LON",
    "service": "6",
    "service_version": "0",
    "service_environment": "test",
    "path": "/dev/sda1",
    "fstype": "ext3",
    "_value": 11.12,
    "metric_name": "cpu.usr"
}
```
https://docs.splunk.com/Documentation/Splunk/7.0.0/Metrics/GetMetricsInOther

Splunk 8.x will still interpret the 7.x schema - I believe the motivation for the 8.x schema is to allow multiple metrics to be added to the same event.

I'm inclined to match the version 8.x schema, but am happy to add support for the version 7.x schema if support for Splunk 7.x is desirable. I believe it is due to go out of support in October. What are your thoughts?

**Comment by jszwedko** [comment]
Ah, thanks @chris1786 . My preference would be to support the 7.x schema if 8.x can read it as well. We can update to 8.x later once 7.x support is dropped.

**Comment by chris1786** [comment]
Thanks @jszwedko. In that case, I'll implement the 7.x schema

**Comment by AdAndre** [comment]
I would also be very interested in using the HEC sink for metrics 👍 

As you wrote, it is already possible to send metrics to a Splunk metric index with the vector http sink after converting them into logs.
In my company, we use vector to convert the metrics to logs and modify the timestamp format, but we use an internally developed tool to convert the metrics to the Splunk format.

Before :
{"gauge":{"value":973.0},"kind":"absolute","name":"scrape_samples_scraped","tags":{"environment":"data-dataapi-01.data.eventhub.adista.fr","instance":"splunk-splsh-01","job":"node_exporter"},"timestamp":1625125400.359}

After :
{"time":1625125400.359,"source":"prometheus://data-dataapi-01.data.eventhub.adista.fr","sourcetype":"hec_metrics","host":"splunk-splsh-01","event":"metric","fields":{"job":"node_exporter","metric_name:scrape_samples_scraped":973.0}}

> 3. Use `remap` to munge the data into the format required by the HEC
> 
> While this works, it is somewhat clunky and unintuitive.

Could you share how you would transform the metric with vector ?
I tried to do it with a remap but I don't manage to get the Splunk 8.x schema for metrics, especially the part with the metric_name inside the field name : "metric_name:cpu.usr": 11.12

My main problem is that I don't know how to use a field value as a field name with vector. Do you know if there is a way of doing it ?
I tried to get around that by using the parse_key_value() function, but then the metric value field becomes a String, and I can't convert it to an Int because the field name will not always be the same...
This is my last try :

.time = del(.timestamp)
.source = del(.tags.environment)
.sourcetype = "hec_metrics"
.host = del(.tags.instance)
.event = "metric"
.fields = merge!(del(.tags),parse_key_value!("metric_name:" + to_string!(.name) + "=" + to_string!(.gauge.value),"="))
del(.kind)
del(.gauge)

-->

{ "event": "metric", "fields": { "job": "node_exporter", "metric_name:scrape_samples_scraped": **_"_973_"_** }, "host": "splunk-splsh-01", "name": "scrape_samples_scraped", "source": "data-dataapi-01.data.eventhub.adista.fr", "sourcetype": "hec_metrics", "time": 1625125400.359}

Thank you for your help

**Comment by chris1786** [comment]
@AdAndre - I've only managed to get the remap working for metrics in the 7.X format. I didn't manage to get the 8.X format, although I didn't spend much time trying. I don't currently have a requirement to push multiple metrics per event, and as Splunk 8.X is compatible with the 7.X format, this is sufficient for my needs for the time being.

I've forked the repo, and written most of the implementation to alter the splunk_hec sink in the following branch:

https://github.com/chris1786/vector/tree/splunk-hec-sink-metrics

I need to update the docs, and have a few bits that need tidying up, but am hoping to raise a PR for this over the next couple of weeks. My plan (at least initially), is to support the 7.X format.

**Comment by zsherman** [comment]
@chris1786 any update here? A PR would be very welcome if you have a working fork.

**Comment by chris1786** [comment]
@zsherman - I still have a few unit tests to write, and need to update the docs. I also need to ensure the implementation caters for namespaces.

I have not had much time to dedicate to this over the past month, but I do intend to complete this and raise a PR over the coming weeks.

**Comment by zsherman** [comment]
@chris1786 got it, let us know if we can assist 👍

**Comment by chris1786** [comment]
PR for initial implementation of this: https://github.com/timberio/vector/pull/8947

**Comment by chris1786** [comment]
Resolved by #8947
